pglock/core.py

```diff
@@ -0,0 +1,394 @@
+            ``pglock.ACCESS_SHARE``.
+        wait (bool, default=True): wait (bool, default=True): If ``True``, will
+            wait to acquire the lock. If ``False``, will return ``True`` if acquired.
+            Runtime behavior can be overridden with the ``side_effect`` argument.
+        using (str, default="default"): The database to use.
+        timeout (datetime.timedelta, default=None): Set a timeout when waiting for the lock.
+            Only significant when using ``wait=True``. Use ``pglock.timedelta`` as a shortcut to
+            avoid importing Python's timedelta object.
+        side_effect (str, default=pglock.RETURN): Adjust the runtime behavior.
+            ``pglock.RETURN`` will return the acquisition status. ``pglock.RAISE`` will
+            raise a ``django.db.utils.OperationalError`` if the lock cannot
+            be acquired or a timeout happens.
+
+    Returns:
+        bool: When using the default side effect, returns ``True`` if the lock was acquired or
+        ``False`` if not.
+
+    Raises:
+        django.db.utils.OperationalError: If ``side_effect=pglock.Raise`` and a lock cannot be
+            acquired or a timeout occurs.
+        RuntimeError: When running code outside of a transaction.
+        ValueError: When ``side_effect`` is an invalid value or no models are supplied.
+
+    """  # noqa
+    if side_effect not in (RETURN, RAISE):
+        raise ValueError("side_effect must be one of pglock.RETURN or pglock.RAISE")
+
+    if not connections[using].in_atomic_block:
+        raise RuntimeError(f'Database "{using}" must be in a transaction to lock models.')
+
+    if not models:
+        raise ValueError("Must supply at least one model to pglock.model().")
+
+    models = [apps.get_model(model) if isinstance(model, str) else model for model in models]
+    models = ", ".join(f'"{model._meta.db_table}"' for model in models)
+    sql = f'LOCK TABLE {models} IN {mode} MODE {"NOWAIT" if not wait else ""}'
+
+    try:
+        with contextlib.ExitStack() as stack:
+            if side_effect == RETURN:
+                # If returning True/False, create a savepoint so that
+                # the transaction isn't in an errored state when returning.
+                stack.enter_context(transaction.atomic(using=using))
+
+            if timeout:
+                stack.enter_context(lock_timeout(timeout, using=using))
+
+            with connections[using].cursor() as cursor:
+                cursor.execute(sql)
+                return True
+    except OperationalError:
+        if side_effect == RETURN:
+            return False
+        elif side_effect == RAISE:
+            raise
+        else:
+            raise AssertionError
+
+
+def _bypass_bg_task(backend_pid, retries, side_effect):
+    BlockedPGLock = apps.get_model("pglock.BlockedPGLock")
+
+    for _ in range(retries + 1):
+        qset = BlockedPGLock.objects.pid(backend_pid)
```

Comment:
I need a bit of reminding on what exactly this is doing here.
Is backend_pid the id of the postgres connection being used by the function wrapped in `bypass`? If so, that means `qset = BlockedPGLock.objects.pid(backend_pid)` fetches everything _blocking_ that connection? So in the case of migrations it would only kill exactly those locks which are blocking the migration, and no more?

Reply:
Additional question about the pg internals of this. Again let's take the case of a migration wrapped in bypass. Is it the case that once the migration has actually acquired the necessary locks, _bypass_ is no longer terminating any pids, since the migration isn't at that point the one blocking, but rather the blocker? If that's the case, we might want to add logic here to detect when our process has acquired locks and stop the periodic nature of bypass at that point.

Reply:
> Additional question about the pg internals of this. Again let's take the case of a migration wrapped in bypass. Is it the case that once the migration has actually acquired the necessary locks, bypass is no longer terminating any pids, since the migration isn't at that point the one blocking, but rather the blocker? If that's the case, we might want to add logic here to detect when our process has acquired locks and stop the periodic nature of bypass at that point.

It's not the case because there could be multiple operations that get blocked later in code wrapped by `bypass`.

Let's say that a migration looks like this:

```
operations = [
    migrations.AlterField("model_one", field="field_one"),
    migrations.AlterField("model_two", field="field_two"),
    migrations.AlterField("model_three", field="field_three"),
]
```

The first operation may be blocked because of the exclusive access lock on model_one. `bypass` eventually kills the blocking lock.

Then the migration moves to the second operation, which also might encounter blocking locks. So we need to keep running bypass until the end of the wrapped code.

Reply:
> Additional question about the pg internals of this. Again let's take the case of a migration wrapped in bypass. Is it the case that once the migration has actually acquired the necessary locks, bypass is no longer terminating any pids, since the migration isn't at that point the one blocking, but rather the blocker? If that's the case, we might want to add logic here to detect when our process has acquired locks and stop the periodic nature of bypass at that point.

> This also makes me think about the "reverse" behaviour of bypass. As I mentioned in the pgmigrate review, we often get the in prod scenario where the blocks coming from the migration after it has acquired the lock are just as damaging as the blocks preventing it acquiring the lock. Could bypass have an option that once the lock is acquired, it then starts killing PIDs which it itself is blocking?

> Alternatively, if we knew these blocking queries came from certain functions, we could decorate them and tell them not to wait for certain other functions and/or locks. Rough ideas:

```
@pglock.terminate_on('migrate')     # could take a list of function paths  or django commands for example
def my_slow_update_func():
       ...
```

OR

```
@pglock.terminate_on(pglock.ACCESS_EXCLUSIVE) 
def my_slow_update_func():
       ...
```

> The first one I'm not exactly sure how we'd do, but it's the most appealing. The second one above would already be good for avoiding clashes with migrations I think.

For the first one, I think you can better accomplish this with advisory locks. I.e take out a "shared" advisory lock on "migrate". When the migrate command runs, it will take out the "exclusive" advisory lock. If it can't take the exclusive lock, it can have a background worker that kills anything that has the shared lock. I.e. the exclusive lock always gets priority.

I.e. you could add a `prioritize` argument to `pglock.advisory` that would accomplish this.

I think it's possible to do the "reverse" behavior like you mentioned. It's just a matter of querying the `blocking_activity_id` field of `BlockedPGLock`. It is a more expensive query though since Postgres doesn't have the inverse of the `pg_blocking_pids` function